nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -484,6 +551,171 @@ impl DataStore {
         Ok(updated)
     }
 
+    /// Updates an instance record by setting the instance's migration ID.
+    //
+    // TODO-design It's tempting to return the updated state of the Instance
+    // here because it's convenient for consumers and by using a RETURNING
+    // clause, we could ensure that the "update" and "fetch" are atomic.
+    // But in the unusual case that we _don't_ update the row because our
+    // update is older than the one in the database, we would have to fetch
+    // the current state explicitly.  For now, we'll just require consumers
+    // to explicitly fetch the state if they want that.
+    pub async fn instance_set_migration_ids(
+        &self,
+        opctx: &OpContext,
+        instance_id: InstanceUuid,
+        src_propolis_id: PropolisUuid,
+        migration_id: Uuid,
+        target_propolis_id: PropolisUuid,
+    ) -> Result<bool, Error> {
+        use db::schema::instance::dsl;
+
+        let instance_id = instance_id.into_untyped_uuid();
+        let target_propolis_id = target_propolis_id.into_untyped_uuid();
+        let src_propolis_id = src_propolis_id.into_untyped_uuid();
+        let updated = diesel::update(dsl::instance)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(instance_id))
+            .filter(dsl::migration_id.is_null())
+            .filter(dsl::target_propolis_id.is_null())
+            .filter(dsl::active_propolis_id.eq(src_propolis_id))
+            .set((
+                dsl::migration_id.eq(Some(migration_id)),
+                dsl::target_propolis_id.eq(Some(target_propolis_id)),
+                // advance the generation
```

Comment:
I think moving this up to Nexus may have introduced a subtle race that can prevent `sis_move_to_starting_undo` from applying an update it wants to apply:

- instance has generation 10
- start saga begins running and installs a Propolis ID in the instance record, moving its generation to 11
- migration saga starts for the same instance, installs migration IDs, sets generation to 12
- start attempt fails and begins to unwind; this will try to set the generation to 12 again, which will fail

This wasn't a problem previously because the migration saga, by virtue of having to call to the instance's current sled to set migration IDs, was prevented from proceeding alongside an in-doubt start saga. That is, if an instance had a running start saga that could still unwind, then it was impossible for a migration saga to operate on the same instance, because by definition there could be no sled agent that would accept a "set migration IDs" request.

That said: Failure to update the instance record in start saga unwind is not *technically* fatal to the saga, in the sense that it won't produce a fatal error that blocks unwinding (we just log a message and continue unwinding). This will end up putting the installed VMM into the `SagaUnwound` state; will a subsequent instance update saga end up cleaning that up correctly?

Reply:
A couple other thoughts (n.b. I haven't worked through these all the way yet):

I'm wondering more and more whether Nexus needs to be doing more with `SagaUnwound` than it is currently. The idea of introducing that state was to make it possible for a start or migrate attempt to unwind without actually having to bump the instance generation number while still allowing a subsequent start/migrate request to proceed (by allowing those requests to overwrite a Propolis/migration ID if it pointed to a an "unwound" record). Does this seem right to you as well? If so I can file an issue.

I also wonder if the migration saga should explicitly check to see if the source VMM has moved out of `Starting` before trying to proceed with the migration. If it's not at least in `Running`, it's possible that the target VMM will start up and find that there isn't yet a source to migrate from.

Reply:
Hm, yeah, I think doing a conditional update here makes sense. 

I wonder if it would make sense to have the migration saga just wait until a start saga completes, rather than failing... but, that seems like a decently large additional complexity that I'm going to punt on for now.

Reply:
Okay, I'm still trying to figure out how to get Diesel to let me do this.

Reply:
f14d3b251022cedf2937a990b0a951dab9f0e12c and 2a60fac0db7e5186b4d2b1a09f86e052fbf7b8dc implemented this!